### PR TITLE
Replace Armadillo with Eigen

### PR DESCRIPTION
**Pull request by hoytak**
Eigen provides benefits over Armadillo in terms of speed, stability, and
    portability.  This PR replaces Armadillo with Eigen 3.  All
    functionality should be essentially uneffected.

### CODE REVIEW

File: src/external/Eigen/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,3 @@
+project(TuriExternalDependencies)
+
+make_empty_library(eigen)
```
Comment (znation) [inline src/external/Eigen/CMakeLists.txt]: Why is it empty? Is Eigen a header-only library? If so, you may not even need a target for it at all.

File: src/unity/extensions/CMakeLists.txt [hunk 2]
```diff
@@ -47,4 +48,11 @@ make_extension(dictionary_transformation_extensions SOURCES
 # )
 # target_link_libraries(timeseries PUBLIC grouped_sframe)
 
+#----------------------------
```
Comment (znation) [inline src/unity/extensions/CMakeLists.txt]: Where did this come from? Doesn't look related to this change.

File: src/ml_data/column_statistics.cpp [hunk 3]
```diff
@@ -237,11 +237,8 @@ void column_statistics::_finalize_global(
             // Adjust for many of the elements being zero
             sa.var_sum += std::pow(sa.mean, 2) * count * (1 - scale);
           }
-          if (total_row_count > 1) {
-            s.stdev = std::sqrt(sa.var_sum / (total_row_count - 1));
-          } else {
-            s.stdev = 0;
-          }
+
+          s.stdev = std::sqrt(sa.var_sum / (total_row_count - 1));
```
Comment (znation) [inline src/ml_data/column_statistics.cpp]: Is this change intentional? It doesn't look related to Eigen.

Comment (hoytak) [reply src/ml_data/column_statistics.cpp]: Good catch.

File: src/unity/extensions/CMakeLists.txt (hunk 2 above)
Comment (hoytak) [reply src/unity/extensions/CMakeLists.txt]: Good catch.

File: src/external/Eigen/CMakeLists.txt (hunk 1 above)
Comment (hoytak) [reply src/external/Eigen/CMakeLists.txt]: True.  Earlier versions had a small configuration step; that is gone now.  IMO it's better to declare the dependency explicitly if possible.  That makes it easier to, e.g. move the package to the deps/ folder.

Comment (znation) [reply src/external/Eigen/CMakeLists.txt]: 🤷‍♂️ 

Personally I don't see why we would want to move this to the deps folder, or why we want anything to explicitly depend on an empty library, but it doesn't really hurt either.

Note that in #1403 I'm actually removing most of the existing empty libraries, as that change splits compile and link-time dependencies apart, and the empty libraries are neither a compile-time nor a link-time dependency of... well, anything, really.

Comment (hoytak) [reply src/external/Eigen/CMakeLists.txt]: I see what you're saying.  I'm okay doing it that way.

**Comment by hoytak** [comment]
> Do you want to delete `src/external/armadillo`?

It's not already deleted?

File: src/serialization/serialize_eigen.hpp [hunk 4]
```diff
@@ -0,0 +1,235 @@
+/**
+ * Copyright (C) 2016 Turi
```
Comment (znation) [inline src/serialization/serialize_eigen.hpp]: Replace this with the Apple copyright header.
